EventVisualisation/Workflow/src/EveWorkflowHelper.cxx

```diff
@@ -137,50 +141,73 @@ void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
       return true;
     }
 
-    TrackTimeNode node;
-    node.trackGID = gid;
-    node.trackTime = static_cast<int>(bracket.mean()) + TIME_OFFSET; // offset the time so it's always positive
+    mGIDTrackTime[gid] = flagTime(bracket.mean(), src);
 
-    // if it's a tracklet, give it a lower priority in time sort by setting the highest bit
-    if (src <= GID::MID) {
-      node.trackTime |= (1 << 31);
+    // If the mode is disabled,
+    // add every track to a symbolic "zero" primary vertex
+    if (!mPrimaryVertexMode) {
+      mPrimaryVertexGIDs[0].push_back(gid);
     }
 
-    trackTimeNodes.push_back(node);
-
     return true;
   };
 
   this->mRecoCont.createTracksVariadic(creator);
 
-  if (trackSorting) {
-    std::sort(trackTimeNodes.begin(), trackTimeNodes.end(),
-              [](TrackTimeNode a, TrackTimeNode b) {
-                return a.trackTime < b.trackTime;
-              });
-  }
+  if (mPrimaryVertexMode) {
+    const auto trackIndex = mRecoCont.getPrimaryVertexMatchedTracks(); // Global ID's for associated tracks
+    const auto vtxRefs = mRecoCont.getPrimaryVertexMatchedTrackRefs(); // references from vertex to these track IDs
+    mTotalPrimaryVertices = vtxRefs.size() - 1;                        // The last entry is for unassigned tracks, ignore them
 
-  std::size_t trackCount = trackTimeNodes.size();
-  if (mEnabledFilters.test(Filter::TotalNTracks) && trackCount >= mMaxNTracks) {
-    trackCount = mMaxNTracks;
-  }
+    for (std::size_t iv = 0; iv < mTotalPrimaryVertices; iv++) {
+      const auto& vtref = vtxRefs[iv];
+      int it = vtref.getFirstEntry(), itLim = it + vtref.getEntries();
+      for (; it < itLim; it++) {
+        const auto tvid = trackIndex[it];
 
-  for (auto node : gsl::span<const TrackTimeNode>(trackTimeNodes.data(), trackCount)) {
-    mTrackSet.trackGID.push_back(node.trackGID);
-    // return the time value to its original form
-    const auto origTime = static_cast<float>(static_cast<int>(node.trackTime & ~(1 << 31)) - TIME_OFFSET);
-    mTrackSet.trackTime.push_back(origTime);
+        if (!mRecoCont.isTrackSourceLoaded(tvid.getSource())) {
+          continue;
+        }
+
+        // TODO: fix TPC tracks?
+
+        // If a track was not rejected, associate it with its primary vertex
+        if (mGIDTrackTime.find(tvid) != mGIDTrackTime.end()) {
```

Comment:
You can spare lot of map find queries, which are not for free, by doing this only for ambiguous tracks
```
if (tvid.isAmbiguous() && mGIDTrackTime.find(tvid) != mGIDTrackTime.end()) { 
  ...
```
let me know if you want to fix this or you want me to merge it as is.